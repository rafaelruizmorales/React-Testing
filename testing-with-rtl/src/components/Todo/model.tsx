export type todoModel = {
    id: number
    title: string
    completed: boolean
}