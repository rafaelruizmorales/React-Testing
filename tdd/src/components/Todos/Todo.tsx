import * as React from 'react'

import { todo } from './model'

interface OwnProps { 
  todo: todo
}

const Todo: React.FC<OwnProps> = (props: OwnProps) => {

  const { id, title, completed } = props.todo

  
  const text = completed 
    ? <><strong>{title}</strong></>
    : <>{title}</>

  return (
    <div data-testid={`todo-${id}`}>
      {text}
    </div>
  )
}

export default Todo