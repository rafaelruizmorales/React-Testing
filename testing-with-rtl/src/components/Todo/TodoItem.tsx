import { todoModel } from "./model";

type todoProps = {
    todo: todoModel;
}

const TodoItem: React.FC<todoProps> = ({todo : {id, title, completed}}) => {

    const text = completed ? <span className="line-through">{title}</span> : <span>{title}</span>

    return (
      <li data-testid={`todo-${id}`}>
        {text}
      </li>
    )
};

export default TodoItem;