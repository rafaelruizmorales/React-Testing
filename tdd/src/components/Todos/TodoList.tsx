import * as React from 'react'

import Todo from './Todo'

interface OwnProps { }

const TodoList: React.FC<OwnProps> = () => {

  const myTodoList = [
    { id: 1, title: 'Learn Testing', completed: false },
    { id: 2, title: 'Take on the World', completed: true }
  ]

  return (
    <div data-testid="todoList_component_wrapper">
      <h1>To Do List</h1>
      {
        myTodoList.map((todo) => {
          return (
            <Todo key={todo.id} todo={todo}/>
          )
        })
      }
    </div>
  )
}

export default TodoList
