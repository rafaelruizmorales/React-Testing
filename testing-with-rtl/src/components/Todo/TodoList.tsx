import { todoModel } from "./model";

import TodoItem from "./TodoItem";

type todoListProps = {
    todos: todoModel[];
}

const TodoList: React.FC<todoListProps> = ({todos}) => {
    return (
        <div data-testid='todoList_component_wrapper'>
            <ul>
                {todos.map(todo => {
                    return <TodoItem key={todo.id} todo={todo} />
                })}
            </ul>
        </div>
    );
};

export default TodoList;