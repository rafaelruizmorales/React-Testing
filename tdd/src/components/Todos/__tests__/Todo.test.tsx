import * as React from 'react';

import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer'

import TodoList from '../TodoList'
import Todo from '../Todo'

describe('Todos', () => {

  afterEach(() => {
    cleanup()
  });

  describe('TodoList Component', () => {

    it('should render TodoList Component', async () => {  
      render(<TodoList />);
      const mainDiv = await screen.findByTestId('todoList_component_wrapper')
      
      expect(mainDiv).toBeInTheDocument()
    });

  });

  describe('Todo Component', () => {

    it('should render NON-completed Todo Component', async () => {
      const todoItem = { id: 1, title: 'Be the best', completed: false }

      render(<Todo key={todoItem.id} todo={todoItem} />);
      const todo = await screen.findByTestId('todo-1')

      expect(todo).toBeInTheDocument()
      expect(todo).toHaveTextContent('Be the best')
      expect(todo).not.toContainHTML('<div data-testid="todo-2"><strong>Have lunch</strong></div>')
    })

    it('should render completed Todo Component', async () => {
      const todoItem = { id: 2, title: 'Have lunch', completed: true }

      render(<Todo key={todoItem.id} todo={todoItem} />);
      const todo = await screen.findByTestId('todo-2')

      expect(todo).toBeInTheDocument()
      expect(todo).toHaveTextContent('Have lunch')
      expect(todo).toContainHTML('<div data-testid="todo-2"><strong>Have lunch</strong></div>')
    })

    it('matches snapshot', () => {
      const todoItem = { id: 1, title: 'Go to bed', completed: true }
      const tree = renderer.create(<Todo key={todoItem.id} todo={todoItem} />).toJSON()

      expect(tree).toMatchSnapshot()

    });

  });
});
