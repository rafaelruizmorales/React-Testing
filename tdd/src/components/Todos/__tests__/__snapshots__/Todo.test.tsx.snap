// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Todo Component matches snapshot 1`] = `
<div
  data-testid="todo-1"
>
  <strong>
    Go to bed
  </strong>
</div>
`;
